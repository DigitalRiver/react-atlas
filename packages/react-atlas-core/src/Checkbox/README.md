Checkbox:

    <Checkbox
	    label="Checkbox"
	  />

Checked:	  

      <Checkbox
        label="Checkbox Checked"
        checked
      />

Disabled:

      <Checkbox
        disabled
        label="Checkbox Disabled"
      />

Left label:

      <Checkbox
        label="Checkbox Left Label"
        labelPosition="left"
        checked
      />

Custom onClick method:

      {}

      <Checkbox
        label="Checkbox onClick"
        value="Bat"
        onClick={function(value, event, isValid, checked) { console.log(value, event, isValid, checked); }}
      />

onChange callback:

    <Checkbox
      label="Checkbox onChange"
      value="Duck"
      onChange={function(value, event, isValid, checked) { console.log(value, event, isValid, checked); }}
      />

No label:

      <Checkbox
      />

No label and checked:

      <Checkbox
        checked
      />

Required:

	  <Checkbox
	  	label="Required Example"
	  	required
	  />

Custom validation:

	  <Checkbox
	  	label="Required Example"
	  	errorCallback={ function(event, checked){ return {valid: checked, message: "Checkbox is required"} } }
	  />

Custom function before change:

      <Checkbox
      	label="Checkbox onBeforeChange"
      	onBeforeChange={ function(value){ if (!value) { let accept = confirm("Do you want to check this?"); return accept; } return true; } }
      />
      
Label and ID:	  

    <Checkbox
      label="Checkbox"
      id="my-checkbox"
    />
