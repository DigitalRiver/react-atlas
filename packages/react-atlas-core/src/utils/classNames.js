export default function classNames(...options) {
  let cn = [];
  options.forEach(opt => {
    if(opt) {
      if(typeof opt === 'string') { 
        cn.push(opt); 
      }
      if(typeof opt === 'object') { 
        let keys = Object.keys(opt).filter(key => !!opt[key]);
        cn.push(...keys);
      }
    }
  });
  return cn;
};