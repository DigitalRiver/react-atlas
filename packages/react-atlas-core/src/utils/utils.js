export default {
  angleFromPositions(cx, cy, ex, ey) {
    const theta = Math.atan2(ey - cy, ex - cx) + Math.PI / 2;
    return theta * 180 / Math.PI;
  },
  angle360FromPositions(cx, cy, ex, ey) {
    const angle = this.angleFromPositions(cx, cy, ex, ey);
    return angle < 0 ? 360 + angle : angle;
  },
  range(start = 0, stop = null, step = 1) {
    let [_start, _stop] = [0, start];
    if (stop !== null) {
      [_start, _stop] = [start, stop];
    }
    const length = Math.max(Math.ceil((_stop - _start) / step), 0);
    const range = Array(length);

    for (let idx = 0; idx < length; idx++, _start += step) {
      range[idx] = _start;
    }

    return range;
  },
  round(number, decimals) {
    if (!isNaN(parseFloat(number)) && isFinite(number)) {
      const decimalPower = Math.pow(10, decimals);
      return Math.round(parseFloat(number) * decimalPower) / decimalPower;
    }
    return NaN;
  },
  getViewport() {
    return {
      "height": window.innerHeight || document.documentElement.offsetHeight,
      "width": window.innerWidth || document.documentElement.offsetWidth
    };
  },
  cloneObject(object) {
    return JSON.parse(JSON.stringify(object));
  },
  inputTypeForPrototype(prototype) {
    if (prototype === Date) {
      return "date";
    }
    if (prototype === Number) {
      return "number";
    }
    if (prototype === Boolean) {
      return "checkbox";
    }
    return "text";
  },
  prepareValueForInput(value, type) {
    if (type === "date") {
      return new Date(value).toISOString().slice(0, 10);
    }
    if (type === "checkbox") {
      return value ? "on" : null;
    }
    return value;
  },
  getSelection(elem) {
    let start, end, rangeElem, clone;

    if (elem.selectionStart !== undefined) {
      start = elem.selectionStart;
      end = elem.selectionEnd;
    }
    else {
      try {
        elem.focus();
        rangeElem = elem.createTextRange();
        clone = rangeElem.duplicate();

        rangeElem.moveToBookmark(document.selection.createRange().getBookmark());
        clone.setEndPoint('EndToStart', rangeElem);

        start = clone.text.length;
        end = start + rangeElem.text.length;
      }
      catch (e) { /* not focused or not visible */ }
    }

    return { start, end };
  },
  setSelection(elem, selection) {
    let rangeElem;

    try {
      if (elem.selectionStart !== undefined) {
        elem.focus();
        elem.setSelectionRange(selection.start, selection.end);
      }
      else {
        elem.focus();
        rangeElem = elem.createTextRange();
        rangeElem.collapse(true);
        rangeElem.moveStart('character', selection.start);
        rangeElem.moveEnd('character', selection.end - selection.start);
        rangeElem.select();
      }
    }
    catch (e) { /* not focused or not visible */ }
  }
};
