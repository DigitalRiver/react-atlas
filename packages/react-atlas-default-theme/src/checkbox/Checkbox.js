import { Checkbox } from "react-atlas-core";
import styles from "./checkbox.css";
import createComponent from "../createComponent";

export default createComponent(Checkbox, styles);
