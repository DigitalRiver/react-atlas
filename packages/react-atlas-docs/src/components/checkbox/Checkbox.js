import React, { PropTypes } from "react";
import * as Atlas from 'react-atlas';

export default class Checkbox extends React.Component {
  constructor(props){
    super(props)
  }
  render() {
    return (
      <Atlas.Checkbox {...this.props}></Atlas.Checkbox>
    )
  }
}