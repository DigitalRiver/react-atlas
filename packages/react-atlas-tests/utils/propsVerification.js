export function verifyPropsDefaultValue(component, expectedProps) {
  //Compares each property value of the component with the expected value

  var result = true;
  var partialResult;

  for (var [compPropName, compPropValue] of expectedProps.entries()) {
    partialResult = component.props()[compPropName] == compPropValue;

    if (!partialResult) {
      console.log(
        "[" +
          compPropName +
          "] -> Got: " +
          component.props()[compPropName] +
          " - expected: " +
          compPropValue
      );
      result = false;
    }
  }
  return result;
}
