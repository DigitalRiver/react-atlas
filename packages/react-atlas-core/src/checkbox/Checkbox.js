import React, { Component, PropTypes } from 'react';
import themeable from 'react-themeable';
import { classNames } from '../utils/utils';


/**
 * Simple component for a basic checkbox
 */
const Checkbox = ({title, label, disabled, inline, ...other}) => {
      //const cx = ClassNames.bind(other.theme);
      const theme = themeable(other.theme);
      const componentClasses = classNames({
          "block": !inline,
          inline,
          disabled,
          [className]: className
      });

      title = title ? title : label;

      return (
          <label {...theme(1, ...componentClasses)} title={title}>
              <input {...other} type="checkbox" disabled={disabled} {...theme(2, 'input')} />
              {label && <span {...theme(3, 'label')}>{label}</span>}
          </label>
      );
};

Checkbox.propTypes = {
  /**
   * Text for checkbox label
   * @examples 'Some Label'
   */
  label: PropTypes.string,
	/**
   * A css class name that will be appended to the wrapping <label> element around the <input> and <span> elements.
   */
  className: PropTypes.string,
	/**
   * If included, checkbox is disabled
   * @examples <Checkbox disabled />, <Checkbox disabled={true} />
   */
  disabled: PropTypes.bool,
	/**
   * If included, renders the checkbox and it's label inline, so it can be side-by-side to other content.
   */
  inline: PropTypes.bool,
  /**
   * Text for checkbox label title. (i.e. "alt-text" for checkboxes, useful for accessibility). If not provided, will be label text.
   * @examples 'Some Title'
   */
  title: PropTypes.string,
	/**
   * Defines if checkbox should be checked on load.
   */
  defaultChecked: PropTypes.bool
};

Checkbox.defaultProps = {
  className: '',
  disabled: false,
  inline: false
};

Checkbox.styleguide = {
  category: "Form Components",
  index: "3.2",
  example: `
<section style={{textAlign:"left"}}>
  <h4 style={{marginBottom: "10px"}}>Checkboxes</h4>
  <Checkbox
    defaultChecked={true}
    label="Checked"
  />
  <Checkbox
    defaultChecked={false}
    label="Not checked"

  />
  <Checkbox
    defaultChecked={true}
    label="Disabled checkbox"
    disabled
  />
  <Checkbox
    defaultChecked={false}
    label="Inline checkbox 1"
    inline
  />
  <Checkbox
    defaultChecked={false}
    label="Inline checkbox 2"
    inline
  />
</section>
`
};

export default Checkbox;
