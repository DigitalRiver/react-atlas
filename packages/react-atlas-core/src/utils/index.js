import events from "./events";
import prefixer from "./prefixer";
// import time from './time';
import utils from "./utils";
import classNames from "./classNames";

export { events, prefixer, utils, classNames };
