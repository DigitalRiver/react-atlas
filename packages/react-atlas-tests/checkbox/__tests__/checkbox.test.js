import React from "react";
import { mount } from "enzyme";
import { default as Checkbox } from "../../../react-atlas-core/src/Checkbox/Checkbox";

import { verifyPropsDefaultValue } from "../../utils/propsVerification";

describe("Test checkbox component", () => {
  it("Test default props", function() {
    const result = mount(<Checkbox />);
    expect(result.props().className).toBe("");
    expect(result.props().disabled).toBe(false);
    expect(result.props().inline).toBe(false);
  });

  it("Make sure label is used as title when title is not set", function() {
    const result = mount(<Checkbox label={"labelText"} />);
    expect(result.text()).toContain("labelText");
  });
});

describe("Test checkbox component - Default values", () => {
  it("Test checkbox component - Custom values", function() {
    const component = mount(
      <Checkbox
        defaultChecked={true}
        title="Adrian rules!!"
        label="Disabled checkbox"
      />
    );

    const expProps = new Map([
      ["label", "Disabled checkbox"],
      ["className", ""],
      ["disabled", false],
      ["inline", false],
      ["title", "Adrian rules!!"]
    ]);
    expect(verifyPropsDefaultValue(component, expProps)).toEqual(true);
  });

  it("Test default props", function() {
    const component = mount(<Checkbox />);

    const expProps = new Map([
      ["className", ""],
      ["disabled", false],
      ["inline", false]
    ]);
    expect(verifyPropsDefaultValue(component, expProps)).toEqual(true);
  });

  it("Test checkbox component - Click event", function() {
    const comp = mount(<Checkbox />);
    expect(comp.state().checked).toEqual(false);
    comp.simulate("click");
    expect(comp.state().checked).toEqual(true);
  });

  it("Test checkbox component (disabled) - Click event", function() {
    const comp = mount(<Checkbox disabled={true} />);
    expect(comp.state().checked).toEqual(false);
    comp.simulate("click");
    expect(comp.state().checked).toEqual(false);
  });

  it("Test checkbox component - Click event with bocus event first", function() {
    const comp = mount(<Checkbox />);
    expect(comp.state().checked).toEqual(false);
    comp.simulate("focus");
    comp.simulate("click");
    expect(comp.state().checked).toEqual(true);
  });

  it("Test checkbox component - Click event with blur event first", function() {
    const comp = mount(<Checkbox />);
    expect(comp.state().checked).toEqual(false);
    comp.simulate("blur");
    comp.simulate("click");
    expect(comp.state().checked).toEqual(true);
  });

  it("Test checkbox component - Event two Clicks", function() {
    const comp = mount(<Checkbox />);
    expect(comp.state().checked).toEqual(false);
    comp.simulate("click");
    comp.simulate("click");
    expect(comp.state().checked).toEqual(false);
  });

  it("Test checkbox component - Click event (custom onClick callback)", function() {
    const comp = mount(
      <Checkbox
        onClick={function() {
          console.log("onClick callback trigered");
        }}
      />
    );
    expect(comp.state().checked).toEqual(false);
    comp.simulate("click");
    expect(comp.state().checked).toEqual(true);
  });

  it("Test checkbox component - Click event (custom onChange callback)", function() {
    const comp = mount(
      <Checkbox
        onChange={function() {
          console.log("onChange callback trigered");
        }}
      />
    );
    expect(comp.state().checked).toEqual(false);
    comp.simulate("click");
    expect(comp.state().checked).toEqual(true);
  });

  it("Test checkbox component - Click event (custom onBeforeChange)", function() {
    const comp = mount(
      <Checkbox
        onBeforeChange={function() {
          return false;
        }}
      />
    );
    expect(comp.state().checked).toEqual(false);
    comp.simulate("click");
    expect(comp.state().checked).toEqual(true);
  });
});
