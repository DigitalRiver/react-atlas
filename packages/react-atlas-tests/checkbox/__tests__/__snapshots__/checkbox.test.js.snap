// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test checkbox component renders correctly 1`] = `
<div
  className=""
  onClick={[Function]}
  style={undefined}
  styleName="checkbox_padding"
>
  <div
    styleName="inline_block relative padding"
  >
    <label
      htmlFor={undefined}
      styleName="label"
      title="labelText"
    >
      labelText
    </label>
    <div
      styleName="float_left"
    >
      <input
        checked={false}
        className="ra_Input__checkbox ra_styles__marg-b-1 ra_Input__max ra_Input__opacity"
        id={undefined}
        name={undefined}
        onBlur={[Function]}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyPress={[Function]}
        onPaste={[Function]}
        style={undefined}
        type="checkbox"
      />
      <div
        styleName="not_checked"
      />
    </div>
  </div>
</div>
`;
